pkg/controllers/monitoring-stack/controller.go

```diff
@@ -0,0 +1,179 @@
+	split := strings.Split(opts.InstanceSelector, "=")
+	if len(split) != 2 {
+		return fmt.Errorf("invalid InstanceSelector: %s", opts.InstanceSelector)
+	}
+
+	r := &reconciler{
+		k8sClient:             mgr.GetClient(),
+		scheme:                mgr.GetScheme(),
+		logger:                ctrl.Log.WithName("monitoring-stack-operator"),
+		instanceSelectorKey:   split[0],
+		instanceSelectorValue: split[1],
+	}
+
+	// We only want to trigger a reconciliation when the generation
+	// of a child changes. Until we need to update our the status for our own objects,
+	// we can save CPU cycles by avoiding reconciliations triggered by
+	// child status changes.
+	p := predicate.GenerationChangedPredicate{}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&stack.MonitoringStack{}).
+		Owns(&monv1.Prometheus{}).WithEventFilter(p).
+		Owns(&v1.ServiceAccount{}).WithEventFilter(p).
+		Owns(&rbacv1.Role{}).WithEventFilter(p).
+		Owns(&rbacv1.RoleBinding{}).WithEventFilter(p).
+		Owns(&monv1.ServiceMonitor{}).WithEventFilter(p).
+		Complete(r)
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.logger.WithValues("stack", req.NamespacedName)
+	logger.Info("Reconciling monitoring stack")
+	ms, err := r.getStack(ctx, req)
+	if err != nil {
+		// retry since some error has occured
+		return ctrl.Result{}, err
+	}
+	if ms == nil {
+		// no such monitoring stack, so stop here
+		return ctrl.Result{}, nil
+	}
+
+	patchers, err := r.stackComponentPatchers(ms, r.instanceSelectorKey, r.instanceSelectorValue)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, mapper := range patchers {
+		if err := r.reconcileObject(ctx, ms, mapper, logger); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *reconciler) getStack(ctx context.Context, req ctrl.Request) (*stack.MonitoringStack, error) {
+	logger := r.logger.WithValues("stack", req.NamespacedName)
+
+	ms := stack.MonitoringStack{}
+
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, &ms); err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(3).Info("stack could not be found; may be marked for deletion")
+			return nil, nil
+		}
+		logger.Error(err, "failed to get monitoring stack")
+		return nil, err
+	}
+
+	return &ms, nil
+}
+
+func (r *reconciler) reconcileObject(ctx context.Context, ms *stack.MonitoringStack, patcher objectPatcher, logger logr.Logger) error {
+	existing := patcher.empty()
+	logger = logger.WithValues(
+		"GVK", existing.GetObjectKind().GroupVersionKind(),
+		"Name", existing.GetName())
+
+	key := types.NamespacedName{
+		Name:      existing.GetName(),
+		Namespace: existing.GetNamespace(),
+	}
+	err := r.k8sClient.Get(ctx, key, existing)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	createNew := errors.IsNotFound(err)
+
+	desired, err := patcher.patch(existing)
+	if err != nil {
+		return err
+	}
+
+	if ms.Namespace == desired.GetNamespace() {
```

Comment:
why is this check needed?

Reply:
Cross-namespace owner references are disallowed in Kubernetes